### PR TITLE
Easy tracker more models

### PR DESCRIPTION
**Pull request by Slion**
Easy Tracker: Adding support for four and five points models.
Adding basic bad pose filtering.
Auto center feature.

### CODE REVIEW

**Comment by sthalik** [comment]
Another issue is that only 2nd, 4th, and 6th order distortion coefficients are specified. Normally it's not necessary to specify coeffs beyond the 4th one. Please go over the undistort formula and see whether the 6th coeff does anything significant.

**Comment by sthalik** [comment]
Please change the camera API so that `camera::start()` has the signature `std::tuple<bool, info> start()` rather than `bool start(info&)`.

**Comment by sthalik** [comment]
If the camera is meant to provide the focal length, the API should be changed so that the FOV is passed as a param from the user interface. I see that you wrote that code with Kinect in mind but it's still necessary to test it on various user configurations.

File: options/tie.cpp [hunk 1]
```diff
@@ -72,6 +72,13 @@ void tie_setting(value<QVariant>& v, QComboBox* cb)
 
 // XXX TODO need variant with setEnabled based on lambda retval -- sh 20170524
 
+void tie_setting(value<bool>& v, QRadioButton* cb)
+{
+    cb->setChecked(v);
+    value_::connect(cb, SIGNAL(toggled(bool)), &v, SLOT(setValue(bool)), v.DIRECT_CONNTYPE);
```
[the remark this answers is not in the document]
Comment (Slion) [reply options/tie.cpp]: It works fine.

**Comment by Slion** [comment]
> Normally it's not necessary to specify coeffs beyond the 4th one.

Distortion coefficients probably don't matter much in our case but I rather leave them in for the sake of completion. Better safe than sorry.

**Comment by Slion** [comment]
> Please change the camera API so that camera::start() has the signature std::tuple<bool, info> start() rather than bool start(info&).

Actually this is intended as an input and output parameter. It's up to the camera implementation to decide what to do with it. Most OpenCV based camera need proper input of width and height otherwise they fail to start. The Kinect IR camera implementation is not expecting any input and outputs intrinsics.

**Comment by Slion** [comment]
> If the camera is meant to provide the focal length, the API should be changed so that the FOV is passed as a param from the user interface. I see that you wrote that code with Kinect in mind but it's still necessary to test it on various user configurations.

Some of that was answered on my previous comment. Unless I'm missing something the camera API modifications are already on our HEAD. The camera API was modified by the initial Easy Tracker contribution in the backwards compatible way. This pull request is mostly Easy Tracker internals except for the Radio Button Options Tie.

I'm happy for us to further the camera API but to me this is beyond the scope of that pull request. Feel free to open an issue if you want to discuss that.

File: options/tie.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Slion) [reply options/tie.cpp]: > Should the policy be that all camera sources set the focal lengths, at least basing on the FOV?

Ideally yes, however most camera implementation including OpenCV camera implementation do not have that information.

> This is reasonable but right now tracker/easy assumes this for all video sources.

Very true. Easy Tracker will only work properly together with camera implementations providing camera intrinsics.

**Comment by sthalik** [comment]
> Distortion coefficients probably don't matter much in our case but I rather leave them in for the sake 
> of completion. Better safe than sorry.

Then include the 4 distortion coefficients as in

```c++
    float dist_c[4] {};
```

in the camera struct. It doesn't make sense for a generic impl to only allow correcting one particular distortion type. The sixth-order coefficient needs to go unless you can demonstrate its effectiveness (see `cv::undistortPoints`).

**Comment by sthalik** [comment]
Zero-initialize the camera info struct in `tracker-easy.h`:

```c++
video::impl::camera::info iCameraInfo {};
```

**Comment by sthalik** [comment]
Last comment was wrong, it's already zero-initialized

**Comment by sthalik** [comment]
Let's merge it and then I'm making API changes.

**Comment by Slion** [comment]
> The sixth-order coefficient needs to go unless you can demonstrate its effectiveness

It does not harm keeping it. Surely if the Kinect SDK provides it and OpenCV supports it, it makes sense to have it. I would align OpenTrack implementation with the OpenCV implementation which is using 8 kinds of distortion coefficient.

**Comment by sthalik** [comment]
> It does not harm keeping it. Surely if the Kinect SDK provides it and OpenCV supports it, it makes sense to have it. I would align OpenTrack implementation with the OpenCV implementation which is using 8 kinds of distortion coefficient.

Please provide some test results, i.e. delta between `cv::undistortPoints()` and the original positions.

**Comment by Slion** [comment]
> Please provide some test results, i.e. delta between cv::undistortPoints() and the original positions.

It's not worth the effort when all you need to do is use `float dist_c[8] {};` instead of `float dist_c[4] {};`. I would use double precision BTW since the tracker data is using doubles anyway.

I suggest we restart that discussion there https://github.com/opentrack/opentrack/issues/939

**Pull request merged**